Annotate `config.GetValue()` with `[NotNullIfNotNull]`

dahlbyk opened the pull request:
This is annoying:
```csharp
// CS8600: Converting possible null value to non-nullable type.
string foo = config.GetValue("foo", "bar");
```

Before the fix, the updated tests fail to compile:
```text
ConfigurationBinderTests.cs(357,31): error CS8600: Converting null literal or possible null value to non-nullable type.
ConfigurationBinderTests.cs(358,48): error CS8600: Converting null literal or possible null value to non-nullable type.
ConfigurationBinderTests.cs(368,25): error CS8605: Unboxing a possibly null value.
ConfigurationBinderTests.cs(369,29): error CS8605: Unboxing a possibly null value.
ConfigurationBinderTests.cs(370,29): error CS8605: Unboxing a possibly null value.
ConfigurationBinderTests.cs(372,31): error CS8600: Converting null literal or possible null value to non-nullable type.
ConfigurationBinderTests.cs(373,48): error CS8600: Converting null literal or possible null value to non-nullable type.
```

The source generator has been updated as well.

dahlbyk committed:
Annotate GetValue() with [NotNullIfNotNull]

dahlbyk commented:
Hmm, tried adding `System.Diagnostics.CodeAnalysis` to the source generator:
https://github.com/dotnet/runtime/blob/cb978a59424d37958e5e10a1a417afd599c59a5d/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Specs/BindingHelperInfo.cs#L34-L41

But build is failing with:
> error CS0436: The type 'NotNullIfNotNullAttribute' in 'runtime\src\libraries\System.Private.CoreLib\src\System\Diagnostics\CodeAnalysis\NullableAttributes.cs' conflicts with the imported type 'NotNullIfNotNullAttribute' in 'Microsoft.Extensions.Configuration.Binder, Version=9.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. Using the type defined in 'runtime\src\libraries\System.Private.CoreLib\src\System\Diagnostics\CodeAnalysis\NullableAttributes.cs'.

tarekgh commented:
@dahlbyk thanks for submitting the PR. I converted this to a draft PR as there is more need to be done before we can accept this PR. Adding the attribute to these APIs make sense, there is more need to be done here if we want to proceed with that.  

- This is kind of breaking change. you can see the app compatibility failures:
```
❌.packages\microsoft.dotnet.apicompat.task\9.0.100-preview.4.24215.1\build\Microsoft.DotNet.ApiCompat.ValidateAssemblies.Common.targets(16,5): error : API compatibility errors between 'ref/net9.0/Microsoft.Extensions.Configuration.Binder.dll' (left) and 'lib/net9.0/Microsoft.Extensions.Configuration.Binder.dll' (right):
❌.packages\microsoft.dotnet.apicompat.task\9.0.100-preview.4.24215.1\build\Microsoft.DotNet.ApiCompat.ValidateAssemblies.Common.targets(16,5): error CP0016: Cannot add attribute 'System.Diagnostics.CodeAnalysis.NotNullIfNotNullAttribute("defaultValue")' to 'Microsoft.Extensions.Configuration.ConfigurationBinder.GetValue(Microsoft.Extensions.Configuration.IConfiguration, System.Type, string, object?)'.
❌.packages\microsoft.dotnet.apicompat.task\9.0.100-preview.4.24215.1\build\Microsoft.DotNet.ApiCompat.ValidateAssemblies.Common.targets(16,5): error CP0016: Cannot add attribute 'System.Diagnostics.CodeAnalysis.NotNullIfNotNullAttribute("defaultValue")' to 'Microsoft.Extensions.Configuration.ConfigurationBinder.GetValue<T>(Microsoft.Extensions.Configuration.IConfiguration, string, T)'.
```
We need to decide if we should go with the breaking change and file a breaking change doc for it. Also, we'll need to suppress the app compatibility failure:

```
error : API breaking changes found. If those are intentional, the APICompat suppression file can be updated by rebuilding with '/p:ApiCompatGenerateSuppressionFile=true' [D:\a\_work\1\s\src\libraries\Microsoft.Extensions.Configuration.Binder\src\Microsoft.Extensions.Configuration.Binder.csproj]
```

- The attribute `NotNullIfNotNullAttribute` is support only on netstandard 2.1 which the configuration binder and source gen libraries support netstandard 2.0 too. maybe you need to reference the source file https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/NullableAttributes.cs#L82 to fill the missing attribute. 

CC @stephentoub @ericstj if they have any suggestion or recommendation about his change.

dahlbyk commented:
Thanks for the initial feedback! I've conditionally included `NullableAttributes.cs` and suppressed the compatibility failure for `Microsoft.Extensions.Configuration.Binder.dll`.

dahlbyk committed:
Annotate generated GetValue() with [NotNullIfNotNull]

dahlbyk commented:
I took a stab at updating the source generator. Haven't figured out the CS0436 error mentioned [above](https://github.com/dotnet/runtime/pull/101336#issuecomment-2067733988), but I did noticed it's specific to `TargetFramework=net462`.

ericstj commented:
I don't think making a nullable API less nullable should be actually breaking.  I think the issue flagged by APICompat is just that the reference assembly needs to be updated.  Please make sure to update reference source in the ref folder.  We also don't have strongly typed rules for nullability attributes - so those are just going to get flagged no matter what.  If we agree to the change we can suppress the diagnostic.

dahlbyk committed:
Avoid duplicate NullableAttributes.cs

dahlbyk commented on src/libraries/Microsoft.Extensions.Configuration.Binder/src/Microsoft.Extensions.Configuration.Binder.csproj:
```diff
@@ -18,6 +18,10 @@
 
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == 'netstandard2.0'">
+    <Compile Include="$(CoreLibSharedDir)System\Diagnostics\CodeAnalysis\NullableAttributes.cs" Link="System.Private.CoreLib\System\Diagnostics\CodeAnalysis\NullableAttributes.cs" />
+  </ItemGroup>
```
It turns out this is redundant, as `NullableAttributes.cs` is already included in most libraries automatically:
https://github.com/dotnet/runtime/blob/1c824d929b734db420babbc04f44a1491d8a7efb/src/libraries/Directory.Build.targets#L170-L176

This explains https://github.com/dotnet/runtime/pull/101336#issuecomment-2067733988: 
> error CS0436: The type 'NotNullIfNotNullAttribute' in 'src\libraries\System.Private.CoreLib\src\System\Diagnostics\CodeAnalysis\NullableAttributes.cs' conflicts with the imported type 'NotNullIfNotNullAttribute' in 'Microsoft.Extensions.Configuration.Binder, ...'.

`Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests.csproj` has its own `NullableAttributes.cs` that is has ignored, but when actually used it conflicts with `internal` one in `Microsoft.Extensions.Configuration.Binder`:
https://github.com/dotnet/runtime/blob/1c824d929b734db420babbc04f44a1491d8a7efb/src/libraries/Microsoft.Extensions.Configuration.Binder/src/Properties/InternalsVisibleTo.cs#L7

There may be a cleaner way, but I was able to work around this by allowing projects to opt out of the import.

dahlbyk commented on src/libraries/Microsoft.Extensions.Configuration.Binder/src/CompatibilitySuppressions.xml:
Adding `[NotNullIfNotNullAttribute]` to `ref/Microsoft.Extensions.Configuration.Binder.cs` makes this suppression unneccessary.

dahlbyk commented on src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Microsoft.Extensions.Configuration.Binder.SourceGeneration.csproj:
```diff
@@ -18,6 +18,10 @@
     <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Workspaces" Version="$(MicrosoftCodeAnalysisVersion_4_8)" PrivateAssets="all" />
   </ItemGroup>
 
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == 'netstandard2.0'">
+    <EmbeddedResource Include="$(CoreLibSharedDir)System\Diagnostics\CodeAnalysis\NullableAttributes.cs" Link="Resources\NullableAttributes.cs" />
+  </ItemGroup>
```
I fixed my [original CS0436](https://github.com/dotnet/runtime/pull/101336#issuecomment-2067733988), but got a new one.

If I unconditionally inject `NullableAttributes.cs` it causes a build failure:
```text
artifacts\obj\Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests\Debug\net462\generated\Microsoft.Extensions.Configuration.Binder.SourceGeneration\Microsoft.Extensions.Configuration.Binder.SourceGeneration.ConfigurationBindingGenerator\BindingExtensions.g.cs(225,58): error CS0436: The type 'NotNullIfNotNullAttribute' in 'artifacts\obj\Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests\Debug\net462\generated\Microsoft.Extensions.Configuration.Binder.SourceGeneration\Microsoft.Extensions.Configuration.Binder.SourceGeneration.ConfigurationBindingGenerator\NullableAttributes.cs' conflicts with the imported type 'NotNullIfNotNullAttribute' in 'Microsoft.Extensions.Configuration.Binder, Version=9.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. Using the type defined in 'artifacts\obj\Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests\Debug\net462\generated\Microsoft.Extensions.Configuration.Binder.SourceGeneration\Microsoft.Extensions.Configuration.Binder.SourceGeneration.ConfigurationBindingGenerator\NullableAttributes.cs'. [src\libraries\Microsoft.Extensions.Configuration.Binder\tests\SourceGenerationTests\Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests.csproj::TargetFramework=net462]
artifacts\obj\Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests\Debug\net462\generated\Microsoft.Extensions.Configuration.Binder.SourceGeneration\Microsoft.Extensions.Configuration.Binder.SourceGeneration.ConfigurationBindingGenerator\BindingExtensions.g.cs(245,58): error CS0436: The type 'NotNullIfNotNullAttribute' in 'artifacts\obj\Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests\Debug\net462\generated\Microsoft.Extensions.Configuration.Binder.SourceGeneration\Microsoft.Extensions.Configuration.Binder.SourceGeneration.ConfigurationBindingGenerator\NullableAttributes.cs' conflicts with the imported type 'NotNullIfNotNullAttribute' in 'Microsoft.Extensions.Configuration.Binder, Version=9.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. Using the type defined in 'artifacts\obj\Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests\Debug\net462\generated\Microsoft.Extensions.Configuration.Binder.SourceGeneration\Microsoft.Extensions.Configuration.Binder.SourceGeneration.ConfigurationBindingGenerator\NullableAttributes.cs'. [src\libraries\Microsoft.Extensions.Configuration.Binder\tests\SourceGenerationTests\Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests.csproj::TargetFramework=net462]
artifacts\obj\Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests\Debug\net9.0\generated\Microsoft.Extensions.Configuration.Binder.SourceGeneration\Microsoft.Extensions.Configuration.Binder.SourceGeneration.ConfigurationBindingGenerator\BindingExtensions.g.cs(229,58): error CS0436: The type 'NotNullIfNotNullAttribute' in 'artifacts\obj\Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests\Debug\net9.0\generated\Microsoft.Extensions.Configuration.Binder.SourceGeneration\Microsoft.Extensions.Configuration.Binder.SourceGeneration.ConfigurationBindingGenerator\NullableAttributes.cs' conflicts with the imported type 'NotNullIfNotNullAttribute' in 'System.Runtime, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. Using the type defined in 'artifacts\obj\Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests\Debug\net9.0\generated\Microsoft.Extensions.Configuration.Binder.SourceGeneration\Microsoft.Extensions.Configuration.Binder.SourceGeneration.ConfigurationBindingGenerator\NullableAttributes.cs'. [src\libraries\Microsoft.Extensions.Configuration.Binder\tests\SourceGenerationTests\Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests.csproj::TargetFramework=net9.0]
artifacts\obj\Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests\Debug\net9.0\generated\Microsoft.Extensions.Configuration.Binder.SourceGeneration\Microsoft.Extensions.Configuration.Binder.SourceGeneration.ConfigurationBindingGenerator\BindingExtensions.g.cs(249,58): error CS0436: The type 'NotNullIfNotNullAttribute' in 'artifacts\obj\Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests\Debug\net9.0\generated\Microsoft.Extensions.Configuration.Binder.SourceGeneration\Microsoft.Extensions.Configuration.Binder.SourceGeneration.ConfigurationBindingGenerator\NullableAttributes.cs' conflicts with the imported type 'NotNullIfNotNullAttribute' in 'System.Runtime, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. Using the type defined in 'artifacts\obj\Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests\Debug\net9.0\generated\Microsoft.Extensions.Configuration.Binder.SourceGeneration\Microsoft.Extensions.Configuration.Binder.SourceGeneration.ConfigurationBindingGenerator\NullableAttributes.cs'. [src\libraries\Microsoft.Extensions.Configuration.Binder\tests\SourceGenerationTests\Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests.csproj::TargetFramework=net9.0]
```

But without it I get a test failure under `net462`:
```text
    Microsoft.Extensions.SourceGeneration.Configuration.Binder.Tests.ConfigurationBindingGeneratorTests.GetValue_T_Key_DefaultValue [FAIL]
      Generator caused diagnostic in output compilation: 'NotNullIfNotNullAttribute' is inaccessible due to its protection level.
      Stack Trace: ...
    Microsoft.Extensions.SourceGeneration.Configuration.Binder.Tests.ConfigurationBindingGeneratorTests.GetValue_TypeOf_Key_DefaultValue [FAIL]
      Generator caused diagnostic in output compilation: 'NotNullIfNotNullAttribute' is inaccessible due to its protection level.
      Stack Trace: ...
    Microsoft.Extensions.SourceGeneration.Configuration.Binder.Tests.ConfigurationBindingGeneratorTests.GetValue [FAIL]
      Generator caused diagnostic in output compilation: 'NotNullIfNotNullAttribute' is inaccessible due to its protection level.
      Stack Trace: ...
```

Unsure how to proceed.

tarekgh commented:
@dahlbyk We shouldn't enforce the definition of the nullable attribute in the source generator as it could lead to complications and problems. The source generator only needs to verify if the attribute is defined during compilation and utilize it accordingly. If it's not defined, we should refrain from emitting any code that relies on this attribute. Users can choose to define the attribute in their code if they wish to support it, granting them flexibility in its usage.

dahlbyk commented:
> The source generator only needs to verify if the attribute is defined during compilation and utilize it accordingly.

This seems reasonable but I'm new to SG so not sure what this looks like. Any examples/docs?

tarekgh commented:
> This seems reasonable but I'm new to SG so not sure what this looks like. Any examples/docs?

You may look at the code https://github.com/dotnet/runtime/blob/cdfc8f255ddda5f6578e55fabb3b7382e6770730/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Parser/KnownTypeSymbols.cs#L78 as example to see how you can search for a type in the compilation. You may look at the docs [Compilation.GetTypesByMetadataName](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.compilation.gettypesbymetadataname?view=roslyn-dotnet-4.7.0#microsoft-codeanalysis-compilation-gettypesbymetadataname(system-string)) and [Compilation.GetTypeByMetadataName](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.compilation.gettypebymetadataname?view=roslyn-dotnet-4.7.0) too for more info. 

You search for `NotNullIfNotNullAttribute` type then if it is found, utilize it.

dahlbyk commented:
1. Updated to only emit `[NotNullIfNotNull]` if it exists
2. Dropped all the churn from the branch
3. Rebased onto latest `main`

------------------------------------

I'm curious if the build is going to reproduce local failures that seem related to #95830:
> System.TypeLoadException: Method 'get_Keys' in type 'ImplementerOfIDictionaryClass`2' from assembly 'Microsoft.Extensions.Configuration.Binder.Tests, Version=9.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60' does not have an implementation.
>
> System.TypeLoadException: Method 'get_Keys' in type 'ImplementerOfIDictionaryClass`2' from assembly 'Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests, Version=9.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60' does not have an implementation.

It does, in fact, have an implementation: https://github.com/dotnet/runtime/blob/dad01daffc3fa4e257abe8e33cb9f80fbaff71d7/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.TestClasses.Collections.cs#L226

I might just need to do a full rebuild?

tarekgh commented on src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Parser/KnownTypeSymbols.cs:
```diff
@@ -132,6 +133,9 @@ public KnownTypeSymbols(CSharpCompilation compilation)
             Delegate = Compilation.GetSpecialType(SpecialType.System_Delegate);
+
+            // Only generate nullable attributes if available
+            NotNullIfNotNullAttribute = compilation.GetBestTypeByMetadataName(typeof(System.Diagnostics.CodeAnalysis.NotNullIfNotNullAttribute));
```
>typeof(System.Diagnostics.CodeAnalysis.NotNullIfNotNullAttribute) [](http://example.com/codeflow?start=78&length=65)

please convert this to a string `System.Diagnostics.CodeAnalysis.NotNullIfNotNullAttribute` instead of using typeof.

tarekgh commented on src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests.csproj:
```diff
@@ -8,6 +8,9 @@
     <EnableConfigurationBindingGenerator>true</EnableConfigurationBindingGenerator>
+
+    <!-- Avoid conflict with Microsoft.Extensions.Configuration.Binder (via InternalsVisibleTo) -->
+    <NoIncludeNullableAttributes>true</NoIncludeNullableAttributes>
```
Is this still needed? If I understand correctly, you shouldn't get any conflicts now as we are not injecting any code for the nullable attribute definition in the source gen test.

dahlbyk replied:
> Is this still needed?

It is. From https://github.com/dotnet/runtime/pull/101336#discussion_r1575655302:

> `Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests.csproj` has its own `NullableAttributes.cs` that it has ignored, but when actually used it conflicts with `internal` one in `Microsoft.Extensions.Configuration.Binder` [via `[InternalsVisibleTo]`].

tarekgh replied:
The attribute definition is inserted to `Microsoft.Extensions.Configuration.Binder` only with the condition `'$(TargetFrameworkIdentifier)' != '.NETCoreApp'"`.  If the test target is `.NETCoreApp`, then you shouldn't get any conflict because the attribute will not be included in `Microsoft.Extensions.Configuration.Binder`. 
If the test target is netfx, then the test itself will not have the definition of the attribute and I guess shouldn't get a conflict too because the attribute will be defined only in  `Microsoft.Extensions.Configuration.Binder`. did you try to remove `NoIncludeNullableAttributes` and still getting the conflict error?

dahlbyk replied:
I pushed a revert to demonstrate the failure: https://github.com/dotnet/runtime/pull/101336/checks?check_run_id=24337379290.

> artifacts\obj\Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests\Release\net462\generated\Microsoft.Extensions.Configuration.Binder.SourceGeneration\Microsoft.Extensions.Configuration.Binder.SourceGeneration.ConfigurationBindingGenerator\BindingExtensions.g.cs(245,58): error CS0436: (NETCORE_ENGINEERING_TELEMETRY=Build) The type 'NotNullIfNotNullAttribute' in 'D:\a\_work\1\s\src\libraries\System.Private.CoreLib\src\System\Diagnostics\CodeAnalysis\NullableAttributes.cs' conflicts with the imported type 'NotNullIfNotNullAttribute' in 'Microsoft.Extensions.Configuration.Binder, Version=9.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. Using the type defined in 'D:\a\_work\1\s\src\libraries\System.Private.CoreLib\src\System\Diagnostics\CodeAnalysis\NullableAttributes.cs'.

The failure is attributed to lines 225 and 245 of `BindingExtensions.g.cs`, where `[NotNullIfNotNull]` is used. The duplicate `NotNullIfNotNullAttribute` is not actually new (it's been inserted into both `net462` assemblies by `Directory.Build.targets` all along), but it only manifests as CS0436 when it comes time to decide which attribute to use.

So either we can prevent the duplicate (by letting the test project opt out of the include, in favor of the one from MECB), or just ignore CS0436 in the SG test project. The former seems preferable to me.

tarekgh replied:
Sorry for convenience. 

I am seeing after you reverted the `NoIncludeNullableAttributes` we still getting the failure. 

```
##[error]artifacts\obj\Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests\Release\net462\generated\Microsoft.Extensions.Configuration.Binder.SourceGeneration\Microsoft.Extensions.Configuration.Binder.SourceGeneration.ConfigurationBindingGenerator\BindingExtensions.g.cs(225,58): error CS0436: (NETCORE_ENGINEERING_TELEMETRY=Build) The type 'NotNullIfNotNullAttribute' in 'D:\a\_work\1\s\src\libraries\System.Private.CoreLib\src\System\Diagnostics\CodeAnalysis\NullableAttributes.cs' conflicts with the imported type 'NotNullIfNotNullAttribute' in 'Microsoft.Extensions.Configuration.Binder, Version=9.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. Using the type defined in 'D:\a\_work\1\s\src\libraries\System.Private.CoreLib\src\System\Diagnostics\CodeAnalysis\NullableAttributes.cs'.
D:\a\_work\1\s\artifa
```

Please return back `NoIncludeNullableAttributes`. Thanks!

tarekgh replied:
Looks you need to rebase with the main one more time.

tarekgh replied:
My reply crossed yours :-). Yes, now it is clearer why we are getting the failure when not having `NoIncludeNullableAttributes`. I thought the nullable attribute definition is not added to the test projects but this is not the case. So, your original change having `NoIncludeNullableAttributes` make sense now.

CC @ericstj if having any better suggestion.

ericstj replied:
Maybe some change could be made to suppress a warning in local definition when using internals visible to.  IMO it's better to have a local definition than depend on one in another assembly.  It's just tests here, so no big deal but wanted to share the suggestion.

dahlbyk commented:
Restored `NoIncludeNullableAttributes` and rebased onto `main` again.

tarekgh commented on src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/ConfigurationBinder.cs:
```diff
@@ -87,11 +88,22 @@ private void EmitGetValueMethods()
 
+            private void EmitNotNullIfNotNull(string parameterName)
+            {
+                if (!_emitNotNullIfNotNull)
```
nit: 

```C#
            private void EmitNotNullIfNotNull(string parameterName)
            {
                if (_emitNotNullIfNotNull)
                {
                    _writer.WriteLine($"[return: global::System.Diagnostics.CodeAnalysis.NotNullIfNotNull(nameof({parameterName}))]");
                }
            }
```

tarekgh commented on src/libraries/Directory.Build.targets:
```diff
@@ -170,6 +170,7 @@
   <!-- Adds Nullable annotation attributes to C# non .NETCoreApp builds. -->
   <ItemGroup Condition="'$(Nullable)' != '' and
                         '$(Nullable)' != 'disable' and
+                        '$(NoIncludeNullableAttributes)' != 'true' and
```
>NoIncludeNullableAttributes [](http://example.com/codeflow?start=27&length=27)

nit: 

Should we call it `SkipAddingNullableAttributes`? 

CC @ViktorHofer

dahlbyk replied:
Renamed to `SkipIncludeNullableAttributes`.

tarekgh reviewed (approved):
Thanks @dahlbyk, I added a minor comments, otherwise LGTM .

dahlbyk commented on src/libraries/Microsoft.Extensions.Configuration.Binder/Microsoft.Extensions.Configuration.Binder.sln:
```diff
@@ -110,6 +110,10 @@ EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "ref", "ref", "{F43534CA-C419-405E-B239-CDE2BDC703BE}"
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "src", "src", "{94EEF122-C307-4BF0-88FE-263B89B59F9F}"
+	ProjectSection(SolutionItems) = preProject
+		..\Directory.Build.props = ..\Directory.Build.props
+		..\Directory.Build.targets = ..\Directory.Build.targets
+	EndProjectSection
```
Easy to revert, but added these to lend visibility to the props/targets that affect compilation (i.e. searching for `SkipIncludeNullableAttributes` finds all instances).

tarekgh replied:
Please revert all solution changes. Thanks!

tarekgh commented:
Thanks @dahlbyk for your help getting this working and completing it.
